File: Back End/C++/Chapter 7 - Inheritance - Extending Classes/2K230204/inheritance_multilevelInheritance.cpp
// Inheritance

/*
Multilevel Inheritance:

                                All Users
                    --------------|--|--|----------------
                    |             |  |  |               |
                    |    ---------------------------    |
                    |    |                         |  <-------  Own memeber functions
    derived class ---- > |    -----------------  <------------  and friendly functions
    member          |    |    |   private   <-----------------  and classes
    functions     --------->  -----------------    |    |       
                    |    |       protected         |    |
                    |    ---------------------------    |
                    |            public                 |
                    -------------------------------------

        fig. A simple view of access controle to the member of a class


                                ---------
                base class      |   A   |       Grandfather
                                ---------
                                    |
                                    |
                                ---------
                Intermediate    |   B   |       Father
                base class      ---------
                                    |
                                    |
                                ---------
                Derived class   |   C   |       Child
                                ---------

        fig. Multilavel inheritance


    A derived class with multilavel inheritance in declared is declared as follows:

        class A {...};              // base class
        class B : public A {...};   // B derived from A
        class C : public B {...};   // C derived from B

    This process can be extened to any number of levels.
*/

#include<iostream>

// BASE CLASS ---------------------------------------------------------------------
class student
{
    protected:
        int roll_number;
    public:
        void get_number(int);
        void put_number(void);
};

void student::get_number(int a)
{
    roll_number = a;
}

void student::put_number(void)
{
    std::cout << "Roll Number: " << roll_number << "\n";
}


// INTERMEDIATE BASE CLASS --------------------------------------------------------
class test : public student      // First level derivation 
{
    protected:
        float sub1;
        float sub2;
    public:
        void get_marks(float, float);
        void put_marks(void);
};

void test::get_marks(float x, float y)
{
    sub1 = x;
    sub2 = y;
}

void test::put_marks()
{
    std::cout << "Marks in SUB1 = " << sub1 << "\n";
    std::cout << "Marks in SUB2 = " << sub2 << "\n";
}


// DERIVED CLASS ------------------------------------------------------------------

class result : public test      // Second level derivation
{
        float total;
    public:
        void display(void);
};

void result :: display(void)
{
    total = sub1 + sub2;
    put_number();
    put_marks();
    std::cout << "Total = " << total << "\n";
}

// MAIN METHOD ---------------------------------------------------------------------

int main()
{
    result student1;        // student1 created

    student1.get_number(111);
    student1.get_marks(75.0, 59.5);

    student1.display();

    return 0;
}

File: Software/Python/Chapter 10 - Object-oriented programming (OOPs)/Concept 03 - Inheritance/2K230320/inheritance_multilevelInheritance.py
# Object-oriented programming
#
# Inheritance:
#     2.  Multilevel inheritance


# Base class --------------------------------------------------------
class students:
    def __init__(self):
        self.rollnumber = 0

    def get_number(self, rnum):
        self.rollnumber = rnum

    def put_number(self):
        return self.rollnumber


# Intermediate class --------------------------------------------------------
class test(students):  # first level derivation
    def __init__(self):
        self.sub1 = 0
        self.sub2 = 0

    def get_marks(self, sb1, sb2):
        self.sub1 = sb1
        self.sub2 = sb2

    def put_marks(self):
        print("Marks in subject1: ", self.sub1)
        print("Marks in subject2: ", self.sub2)


# Derived class --------------------------------------------------------
class result(test):  # second level derivation
    def __init__(self):
        self.total = 0

    def display(self):
        self.total = self.sub1 + self.sub2

        print(end='\n')
        print("Roll number:", students.put_number(self), end="\n")
        test.put_marks(self)

        print("Roll number:", self.rollnumber, "|", "Total marks:", self.total, end="\n")
        print(end='\n')


# Main Method --------------------------------------------------------

std1 = result()

std1.get_number(10)
std1.get_marks(99, 100)
std1.display()

std2 = result()

std2.get_number(20)
std2.get_marks(98, 99)
std2.display()

print(end="\n")
